### ISSUE TITLE
2 is not a valid CSR version

### ISSUE THREAD

**Issue by rogermartensson** [grindsa/acme2certifier#96]
I'm trying to set up a local ACME server with the mswcce ca_handler and I'm getting an error that I do not know how to debug.

The message is: ```ca_server.get_cert() failed with error: 2 is not a valid CSR version```

I get this error with both master and with version 0.22.

The CSR is created from using certbot. When looking at the created CSR from certbot it says version 3 (0x2) .

**Comment by rogermartensson** [comment]
Just for information. I created a version 1 CSR and used the --csr switch in certbot. This resulted in another type of error:
```ca_server.get_cert() failed with error: RequestSessionError: code: 0x80070057 - E_INVALIDARG - One or more arguments are invalid.```

**Comment by grindsa** [comment]
Can you enable debugging  in acme2certifer (option `debug: True` in `acme_srv.cfg`) and share the logs? 
The certbot version as well as the  command line for enrollment you did use would also be helpful.
Last but not least. Which CA handler are you using?

**Comment by grindsa** [comment]
Another thought coming into my mind: Certbot is not populating the CN field in the CSR. Could it be that your CA is mandating this and the above error message is just misleading? Can you try another acme-client ([acme.sh](https://github.com/acmesh-official/acme.sh) or [lego](https://github.com/go-acme/lego))?

**Comment by rogermartensson** [comment]
We'll look at our PKI and see if there is any rules stopping us.  Our PKI admin couldn't see any failed requests in the log. We'll dig deeper.

The certbort command used is below. We first tested without the --csr switch and as you said it created CSR without a CommonName. We tried to create our own CSR with a CN and got another error.
```
certbot certonly --server http://ouracmeserver.example.org:22280  --standalone --preferred-challenges http -d ourtestwebaddress.example.org --cert-name ourtestwebaddress.exampleorg --csr ourcsr.csr
```

I'll do some more testing and return with debug logs.

**Comment by rogermartensson** [comment]
Certbot version: 0.40.0

I did try to issue a certificate manually with a certbot created CSR using the certsrv web interface and I didn't get any errors.

I tried to use certipy against our CA and it looks like it works.
```
/usr/local/bin/certipy req -username username@addomain.example.org -password WIko1010 -ca "Mid Sweden University CA v2" -target ca.miun.se -template WebServer -debug
Certipy v4.0.0 - by Oliver Lyak (ly4k)

[+] Trying to resolve 'ca.example.org' at '127.0.0.53'
[+] Trying to resolve 'ADDOMAIN.EXAMPLE.ORG' at '127.0.0.53'
[+] Generating RSA key
[*] Requesting certificate via RPC
[+] Trying to connect to endpoint: ncacn_np:a.b.c.d[\pipe\cert]
[+] Connected to endpoint: ncacn_np:a.b.c.d[\pipe\cert]
[*] Successfully requested certificate
[*] Request ID is 47174
[*] Got certificate without identification
[*] Certificate has no object SID
[*] Saved certificate and private key to 'username.pfx'
```

This is a somewhat cleaned up log. Hope this is enough for you to get an idea.
```
acme-srv_1  | [Mon Sep 26 09:31:37.249291 2022] [wsgi:error] [pid 14:tid 140647092713024] CAhandler._config_load()
acme-srv_1  | [Mon Sep 26 09:31:37.249340 2022] [wsgi:error] [pid 14:tid 140647092713024] load_config(/var/www/acme2certifier/acme_srv/acme_srv.cfg:CAhandler)
acme-srv_1  | [Mon Sep 26 09:31:37.249951 2022] [wsgi:error] [pid 14:tid 140647092713024] CAhandler._config_load() ended
acme-srv_1  | [Mon Sep 26 09:31:37.249997 2022] [wsgi:error] [pid 14:tid 140647092713024] Certificate._enroll_and_store(): trigger enrollment
acme-srv_1  | [Mon Sep 26 09:31:37.250033 2022] [wsgi:error] [pid 14:tid 140647092713024] CAhandler.enroll(WebServer)
acme-srv_1  | [Mon Sep 26 09:31:37.250064 2022] [wsgi:error] [pid 14:tid 140647092713024] CAhandler.request_create()
acme-srv_1  | [Mon Sep 26 09:31:37.250352 2022] [wsgi:error] [pid 14:tid 140647092713024] Trying to resolve 'ca.example.org' at '127.0.0.11'
acme-srv_1  | [Mon Sep 26 09:31:37.252429 2022] [wsgi:error] [pid 14:tid 140647092713024] Trying to connect to endpoint: ncacn_np: a.b.c.d[\\pipe\\cert]
acme-srv_1  | [Mon Sep 26 09:31:37.290874 2022] [wsgi:error] [pid 14:tid 140647092713024] Connected to endpoint: ncacn_np:a.b.c.d[\\pipe\\cert]
acme-srv_1  | [Mon Sep 26 09:31:37.303609 2022] [wsgi:error] [pid 14:tid 140647092713024] CAhandler.request_create() ended
acme-srv_1  | [Mon Sep 26 09:31:37.303780 2022] [wsgi:error] [pid 14:tid 140647092713024] build_pem_file()
acme-srv_1  | [Mon Sep 26 09:31:37.304230 2022] [wsgi:error] [pid 14:tid 140647092713024] Requesting start
acme-srv_1  | [Mon Sep 26 09:31:37.304444 2022] [wsgi:error] [pid 14:tid 140647092713024] ca_server.get_cert() failed with error: 2 is not a valid CSR version
acme-srv_1  | [Mon Sep 26 09:31:37.304485 2022] [wsgi:error] [pid 14:tid 140647092713024] cert bundling failed
acme-srv_1  | [Mon Sep 26 09:31:37.304545 2022] [wsgi:error] [pid 14:tid 140647092713024] Certificate.enroll() ended
acme-srv_1  | [Mon Sep 26 09:31:37.304641 2022] [wsgi:error] [pid 14:tid 140647092713024] acme2certifier enrollment error: cert bundling failed
acme-srv_1  | [Mon Sep 26 09:31:37.304747 2022] [wsgi:error] [pid 14:tid 140647092713024] Certificate._enroll_and_store(): invalidating order as there is no certificate and no poll_identifier: cert bundling failed/6XJSW7yhnxNa
```

**Comment by rogermartensson** [comment]
One last entry. I found where certbot create CSR and it is hardcoded for version 2 CSRs.
I changed so certbot issued version 0 CSR and I'm now able to issue  new certificates.

If you want to try it yourself you can change version in make_csr in this file.
https://github.com/certbot/certbot/blob/master/acme/acme/crypto_util.py

**Comment by grindsa** [comment]
Thanks for troubleshooting yourself and for finding the root-cause.  The remark to have an additional check for config-options starting or ending with " makes sense. I will implement this in the next release. 

As for the CSR version issue. I need to look into this. We have a test between certbot and the wcce handler as part of our regression and the combination works without issues. Will keep you updated on this...

**Comment by grindsa** [comment]
ok I think got the full picture now. The error message  `2 is not a valid CSR version` comes from the cryptography library used by acme2certifer.  It has been introduced in [May this year](https://github.com/pyca/cryptography/pull/7249/files#diff-e6131376819316b64b2fc2b73b3cadf57e6169b1315c79dc7ce3748c7c985af7) but made it into a [38.0.0 release in September](https://cryptography.io/en/latest/changelog/).  Certbot already [fixed the issue](https://github.com/certbot/certbot/pull/9334/files#diff-035393341b36c4adff76feb0627156920de618857c156c42335ef9b0437de6fe)
 in [version 1.29](https://github.com/certbot/certbot/releases/tag/v1.29.0) released on July 29th. 

- Our regression did not spot the issue as the fix from certbot came in earlier than the fix in the cryptography module and we do always use the latest versions
- You did observe the problem as you were using an old certbot version with the latest cryptography module.
- I did not see the issue as I did use both old and new certbot versions with an older cryptography (37.x) module.

Learning from all this: we should always use the latest software packages :-)

I will leave the issue open till I implemented the above mentioned config check.
